### PR TITLE
feat: 🎸 wip tabs

### PR DESCRIPTION
**Pull request by shmidt-i**

### CODE REVIEW

File: src/Molecules/NavTabs/NavTabs.tsx [hunk 1]
```diff
@@ -0,0 +1,143 @@
+import React, { createContext, useState, useContext } from 'react';
+import { Route, matchPath, __RouterContext } from 'react-router';
+import { Link } from 'react-router-dom';
+import styled, { css } from 'styled-components';
+import { Flexbox, Typography } from '../../index';
+import { assert } from '../../common/utils';
+
+const useKeyboardNavigation = ({ itemsLength, onChange = (n: number) => null }) => {
```
Comment (JacobBlomgren) [inline src/Molecules/NavTabs/NavTabs.tsx]: this could be broken out since it's duplicated

File: src/Molecules/NavTabs/NavTabs.tsx [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  // const useKeyboardNavigation: Arguments => Return = ({ itemsLength, onChange }) => {
```
Comment (JacobBlomgren) [inline src/Molecules/NavTabs/NavTabs.tsx]: And we need types, this is from my failed attempt at that

File: src/Molecules/NavTabs/NavTabs.tsx [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  const context = useContext(TabContext);
+  assert(
+    typeof context !== 'undefined',
+    `Please don't use <Tabs.Item/> outside of <Tabs.Container />`,
+  );
+  const { active } = context;
+  return <div>{active ? children : null}</div>;
+};
+
+const styles = css`
```
Comment (JacobBlomgren) [inline src/Molecules/NavTabs/NavTabs.tsx]: styles are now duplicated

File: src/Molecules/Tabs/Tabs.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import { storiesOf } from '@storybook/react';
+import React from 'react';
+import Tabs from './Tabs';
+
+storiesOf('Molecules | Tabs', module).add('Default Tabs component', () => (
+  <>
+    <Tabs.Container>
+      <Tabs.Item title="One">Ones children</Tabs.Item>
+      <Tabs.Item title="Node as well 👎">I'm a link asdasdasdads</Tabs.Item>
```
Comment (JacobBlomgren) [inline src/Molecules/Tabs/Tabs.stories.tsx]: debug stuff left over

File: src/Molecules/NavTabs/NavTabs.tsx [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  margin-inline-end: unset;
+  list-style: none;
+  display: flex;
+  /** @todo reconsider spacing */
+  padding-left: ${p => p.theme.spacing.unit(3)}px;
+  /** @todo check this out */
+  margin-bottom: -1px;
+`;
+const isItemElement = (x: any): x is { type: typeof Item; props: ItemProps } =>
+  x != null && typeof x === 'object' && Object.hasOwnProperty.call(x, 'type');
```
Comment (jauniuse) [inline src/Molecules/NavTabs/NavTabs.tsx]: is this check enough? What I have in mind is:

```js
 x.type === Item
```

File: src/Molecules/NavTabs/NavTabs.tsx [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+      titles.push(
+        <Flexbox.Item as="li" role="presentation">
+          <Title active={isIndexActive} setRef={setRef(i)} to={c.props.to} onKeyDown={onKeyDown}>
+            {c.props.title}
+          </Title>
+        </Flexbox.Item>,
+      );
+
+      if (isIndexActive) contents.push(<section>{c}</section>);
```
Comment (jauniuse) [inline src/Molecules/NavTabs/NavTabs.tsx]: should `contents` be array if it's always one active index?

File: src/Molecules/Tabs/Tabs.tsx [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+`;
+
+const StyledButton = styled(NormalizedElements.Button)`
+  ${styles}
+`;
+
+const Title: React.FC<TitleProps> = ({
+  active: activeFromProps,
+  children,
+  onTitleClick,
```
Comment (jauniuse) [inline src/Molecules/Tabs/Tabs.tsx]: I think this prop should be named `onClick`

File: src/Molecules/CardWithTitle/CardWithTitle.tsx [hunk 8]
```diff
@@ -4,11 +4,9 @@ import R from 'ramda';
 import { Typography, Card } from '../..';
 import { CardWithTitleComponent, Props } from './CardWithTitle.types';
 
-const StyledCard = styled(Card)`
-  padding: ${({ theme }) => theme.spacing.unit(5)}px;
-`;
+const StyledCard = styled(Card)``;
```
Comment (shmidt-i) [inline src/Molecules/CardWithTitle/CardWithTitle.tsx]: Just Card then?

File: src/Molecules/TabsNav/TabsNav.tsx [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  margin-top: 0;
+  list-style: none;
+  display: flex;
+  /** @todo reconsider spacing */
+  padding-left: ${p => p.theme.spacing.unit(5)}px;
+  /** @todo check this out */
+  margin-bottom: -1px;
+`;
+const isItemElement = (x: any): x is { type: typeof Item; props: ItemProps } =>
+  x != null && typeof x === 'object' && Object.hasOwnProperty.call(x, 'type');
```
Comment (shmidt-i) [inline src/Molecules/TabsNav/TabsNav.tsx]: Add here type === Item check too

File: src/Molecules/TabsNav/TabsNav.tsx [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+const TabsNav: Component = ({ children }) => {
+  const { location } = useContext(__RouterContext);
+  const { setRef, onKeyDown } = useKeyboardNavigation({
+    itemsLength: React.Children.count(children),
+  });
+  const titles: React.ReactNode[] = [];
+
+  React.Children.forEach(children, (c, i) => {
+    if (!isItemElement(c)) {
+      assert(false, 'Children type of <Tabs.Container> should be only <Tabs.Item>');
```
Comment (shmidt-i) [inline src/Molecules/TabsNav/TabsNav.tsx]: Change error text to be the same as in TabsWithState

File: src/Molecules/TabsNav/TabsNav.tsx [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+import styled from 'styled-components';
+import { Flexbox, Typography, TabTitle } from '../..';
+import { assert } from '../../common/utils';
+import { useKeyboardNavigation } from '../TabsWithState/useKeyboardNavigation';
+import { ItemComponent, ItemProps, TitleComponent, Component } from './TabsNav.types';
+
+const Item: ItemComponent = ({ children }) => {
+  return <div>{children}</div>;
+};
+Item.displayName = 'TabsNav.Item';
```
Comment (shmidt-i) [inline src/Molecules/TabsNav/TabsNav.tsx]: We actually don't render this one?

File: src/Molecules/TabsWithState/TabsWithState.stories.tsx [hunk 12]
```diff
@@ -0,0 +1,37 @@
+import { storiesOf } from '@storybook/react';
+import React from 'react';
+import styled from 'styled-components';
+
+import { TabsWithState, Typography } from '../../index';
+// import Separator from '.Atoms/Separator';
```
Comment (shmidt-i) [inline src/Molecules/TabsWithState/TabsWithState.stories.tsx]: Remove

File: src/Molecules/TabsWithState/TabsWithState.tsx [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+  ItemProps,
+} from './TabsWithState.types';
+
+const Item: ItemComponent = ({ children }) => {
+  return <>{children}</>;
+};
+Item.displayName = 'TabsWithState.Content';
+
+const StyledButton = styled(NormalizedElements.Button)`
+  background: none;
```
Comment (mjmonline) [inline src/Molecules/TabsWithState/TabsWithState.tsx]: this should probably also have `cursor: pointer`

File: src/Atoms/TabTitle/TabTitle.tsx [hunk 14]
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+import styled from 'styled-components';
+import { Props } from './TabTitle.types';
+
+const StyledDiv = styled.div<Props>`
+  background: none;
+  display: inline-block;
+  border: none;
```
Comment (mjmonline) [inline src/Atoms/TabTitle/TabTitle.tsx]: I feel like TabsTitle should be a `<button>` element by default and then we can change that by the `as` prop. It feels weird in storybook to have non interactive tab titles. If we're still rolling with divs then this StyledDiv doesn't need to reset the rules for background, display, border, margin and padding.

File: src/Molecules/TabsWithState/TabsWithState.stories.tsx [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+
+const SpacingInside = styled.div`
+  padding-left: ${p => p.theme.spacing.unit(5)}px;
+  padding-right: ${p => p.theme.spacing.unit(5)}px;
+  padding-top: ${p => p.theme.spacing.unit(4)}px;
+`;
+
+storiesOf('Molecules | TabsWithState', module).add('Default (extra space inside)', () => (
```
Comment (mjmonline) [inline src/Molecules/TabsWithState/TabsWithState.stories.tsx]: Feels like Molecules / tabs would be enough, just my opinion :)

File: src/Molecules/TabsNav/TabsNav.tsx [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+  const titles: React.ReactNode[] = [];
+
+  React.Children.forEach(children, (c, i) => {
+    if (!isItemElement(c)) {
+      assert(false, 'Children type of <Tabs.Container> should be only <Tabs.Item>');
+    } else {
+      const isIndexActive = Boolean(matchPath(location.pathname, c.props.to));
+
+      titles.push(
+        <Flexbox.Item as="li" role="presentation">
```
Comment (mjmonline) [inline src/Molecules/TabsNav/TabsNav.tsx]: Here in the TabsNav component we can remove the `role="presentation"`. It makes more sence in the regular Tabs component

File: src/Molecules/NavTabs/NavTabs.tsx (hunk 5 above)
Comment (shmidt-i) [reply src/Molecules/NavTabs/NavTabs.tsx]: Added

File: src/Molecules/NavTabs/NavTabs.tsx (hunk 6 above)
Comment (shmidt-i) [reply src/Molecules/NavTabs/NavTabs.tsx]: Fixed, no more contents for NavTabs

File: src/Molecules/Tabs/Tabs.tsx (hunk 7 above)
Comment (shmidt-i) [reply src/Molecules/Tabs/Tabs.tsx]: +

File: src/Molecules/CardWithTitle/CardWithTitle.tsx (hunk 8 above)
Comment (shmidt-i) [reply src/Molecules/CardWithTitle/CardWithTitle.tsx]: +

File: src/Molecules/TabsNav/TabsNav.tsx (hunk 9 above)
Comment (shmidt-i) [reply src/Molecules/TabsNav/TabsNav.tsx]: +

File: src/Molecules/TabsNav/TabsNav.tsx (hunk 10 above)
Comment (shmidt-i) [reply src/Molecules/TabsNav/TabsNav.tsx]: +

File: src/Molecules/TabsWithState/TabsWithState.tsx (hunk 13 above)
Comment (shmidt-i) [reply src/Molecules/TabsWithState/TabsWithState.tsx]: +

File: src/Molecules/TabsWithState/TabsWithState.stories.tsx (hunk 15 above)
Comment (shmidt-i) [reply src/Molecules/TabsWithState/TabsWithState.stories.tsx]: +

**Pull request merged**